File: scripts/generate_tags_page.py
#!/usr/bin/env python3
"""ハッシュタグの一覧ページ `tags/index.html` を作る。

記事末尾のハッシュタグの飛び先。タグごとに、そのタグが付いた記事を並べる。
材料は scripts/tags.json（記事→タグ）と news/index.html（記事の日付・タイトル・区分）。

    python3 scripts/generate_tags_page.py
"""
import html
import json
import pathlib
import re
import sys

ROOT = pathlib.Path(__file__).resolve().parents[1]
BASE = "https://www.toyoseeds.com"
ROW = re.compile(
    r'<a class="news-list-row" href="([^"]+)">'
    r'<time datetime="([^"]+)">([^<]+)</time>'
    r'<span class="category([^"]*)"([^>]*)>([^<]*)</span>'
    r"<strong(.*?)>(.*?)</strong></a>",
    re.S,
)

PAGE = """<!doctype html>
<html lang="ja">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<title data-en="Tags | ToyoSeeds LLC">タグ一覧｜ToyoSeeds合同会社</title>
<meta name="description" data-en-content="Every tag on the ToyoSeeds diaries, and the posts under each one." content="ToyoSeedsの日記についているタグと、その記事の一覧です。">
<link rel="canonical" href="{base}/tags/">
<meta property="og:type" content="website">
<meta property="og:title" data-en-content="Tags | ToyoSeeds LLC" content="タグ一覧｜ToyoSeeds合同会社">
<meta property="og:description" data-en-content="Every tag on the ToyoSeeds diaries, and the posts under each one." content="ToyoSeedsの日記についているタグと、その記事の一覧です。">
<meta property="og:url" content="{base}/tags/">
<meta property="og:image" content="{base}/assets/og-image-v3.jpg">
<meta name="twitter:card" content="summary_large_image">
<link rel="stylesheet" href="../assets/content-v2.css">
<script src="../assets/i18n.js" defer></script>
</head>
<body>
<header class="content-header">
  <a class="brand" href="/">Toyo<span>Seeds</span></a>
  <nav aria-label="主要メニュー" data-en-aria-label="Main menu">
    <a href="/" data-en="Home">ホーム</a><a href="/news/">News</a><a href="/#company" data-en="Company">会社概要</a><a class="lang-toggle" href="?lang=en" lang="en" data-lang-toggle aria-label="Switch to English">EN</a><a class="nav-contact" href="/contact/" data-en="Contact">お問い合わせ</a>
  </nav>
</header>
<main class="listing-shell">
  <a class="back-link" href="/news/" data-en="← Diary &amp; News">← 日記・お知らせ</a>
  <p class="eyebrow">TAGS</p>
  <h1 data-en="Tags">タグ一覧</h1>
  <p class="lead" data-en="Every tag on the diaries, and the posts under each one.">日記についているタグと、その記事の一覧です。</p>
  <nav class="tag-cloud" aria-label="タグ" data-en-aria-label="Tags">{cloud}</nav>
{groups}</main>
<footer class="content-footer">
  <a class="brand brand--footer" href="/">Toyo<span>Seeds</span></a>
  <div><a href="/privacy-policy/" data-en="プライバシーポリシー">プライバシーポリシー</a><span>© ToyoSeeds LLC.</span></div>
</footer>
<script>
// i18n.js が英語に差し替えると高さが変わってアンカーがずれるので、差し替え後にもう一度合わせる
addEventListener("load", function () {{
  if (!location.hash) return;
  var el = document.getElementById(decodeURIComponent(location.hash.slice(1)));
  if (el) el.scrollIntoView();
}});
</script>
<script defer src="/_vercel/insights/script.js"></script>
</body>
</html>
"""


def anchor(ja):
    return "tag-" + ja.lstrip("#")


def en_attr(ja, en):
    return f' data-en="{html.escape(en, quote=True)}"' if en and en != ja else ""


def main():
    tags = {k: v for k, v in json.loads((ROOT / "scripts" / "tags.json").read_text(encoding="utf-8")).items()
            if not k.startswith("_")}
    news = (ROOT / "news" / "index.html").read_text(encoding="utf-8")
    rows = {}
    order = []
    for href, iso, dot, cat_mod, cat_attr, cat, t_attr, title in ROW.findall(news):
        slug = href.strip("/")
        rows[slug] = (iso, dot, cat_mod, cat_attr, cat, t_attr, title)
        order.append(slug)

    missing = [s for s in tags if s not in rows]
    if missing:
        sys.exit(f"[中止] news/index.html に見つからない記事: {', '.join(missing)}")

    groups = {}
    for slug in order:                      # news は新しい順。その順のまま各タグに入る
        if slug in tags:
            groups.setdefault(tags[slug]["ja"], []).append(slug)

    # 本数の多いタグから。同数なら新しい記事があるほうを先に
    ordered = sorted(groups, key=lambda t: (-len(groups[t]), order.index(groups[t][0])))

    cloud = "".join(
        f'<a href="#{anchor(t)}"{en_attr(t, tags[groups[t][0]]["en"])}>{html.escape(t)}</a>'
        for t in ordered
    )

    out = []
    for t in ordered:
        en = tags[groups[t][0]]["en"]
        items = "".join(
            '<a class="news-list-row" href="/{s}/"><time datetime="{iso}">{dot}</time>'
            '<span class="category{mod}"{cattr}>{cat}</span><strong{tattr}>{title}</strong></a>'.format(
                s=s, iso=rows[s][0], dot=rows[s][1], mod=rows[s][2], cattr=rows[s][3],
                cat=rows[s][4], tattr=rows[s][5], title=rows[s][6])
            for s in groups[t]
        )
        n = len(groups[t])
        out.append(
            f'  <section class="tag-group" id="{anchor(t)}">\n'
            f'    <h2{en_attr(t, en)}>{html.escape(t)}</h2>\n'
            f'    <p class="tag-count" data-en="{n} post{"s" if n > 1 else ""}">{n}本</p>\n'
            f'    <div class="news-list">{items}</div>\n'
            f"  </section>\n"
        )

    (ROOT / "tags").mkdir(exist_ok=True)
    (ROOT / "tags" / "index.html").write_text(
        PAGE.format(base=BASE, cloud=cloud, groups="".join(out)), encoding="utf-8")
    print(f"tags/index.html を作りました（タグ {len(ordered)}／記事 {len(tags)}）")


if __name__ == "__main__":
    main()

File: scripts/generate_tags_page.mjs
// scripts/generate_tags_page.py の node 移植（トヨのマシンは python が Windows Store のスタブで動かないため）。
// 既存の tags/index.html の head/footer はそのまま残し、<p class="lead"> と </main> の間（タグ並び＋タグごとの一覧）だけ作り直す。
// 材料は scripts/tags.json（記事→タグ）と news/index.html（記事の日付・タイトル・区分）。
//   node scripts/generate_tags_page.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const readLF = (p) => { const s = fs.readFileSync(p, "utf8"); return { text: s.replace(/\r\n/g, "\n"), crlf: s.includes("\r\n") }; };
const writeEOL = (p, text, crlf) => fs.writeFileSync(p, crlf ? text.replace(/\n/g, "\r\n") : text, "utf8");
const esc = (t) => t.replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;").replace(/'/g, "&#x27;");

const ROW = /<a class="news-list-row" href="([^"]+)"><time datetime="([^"]+)">([^<]+)<\/time><span class="category([^"]*)"([^>]*)>([^<]*)<\/span><strong(.*?)>(.*?)<\/strong><\/a>/gs;

const tagsAll = JSON.parse(fs.readFileSync(path.join(ROOT, "scripts", "tags.json"), "utf8"));
const tags = Object.fromEntries(Object.entries(tagsAll).filter(([k]) => !k.startsWith("_")));
const news = readLF(path.join(ROOT, "news", "index.html")).text;

const rows = {};
const order = [];
for (const m of news.matchAll(ROW)) {
  const slug = m[1].replace(/^\/+|\/+$/g, "");
  rows[slug] = { iso: m[2], dot: m[3], mod: m[4], cattr: m[5], cat: m[6], tattr: m[7], title: m[8] };
  order.push(slug);
}
const missing = Object.keys(tags).filter((s) => !rows[s]);
if (missing.length) { console.error(`[中止] news/index.html に見つからない記事: ${missing.join(", ")}`); process.exit(1); }

const groups = {};
const enOf = {};
for (const slug of order) { // news は新しい順。その順のまま各タグに入る
  const list = Array.isArray(tags[slug]) ? tags[slug] : slug in tags ? [tags[slug]] : [];
  for (const tag of list) { (groups[tag.ja] ||= []).push(slug); enOf[tag.ja] = tag.en; }
}
// 本数の多いタグから。同数なら新しい記事があるほうを先に
const ordered = Object.keys(groups).sort((a, b) => (groups[b].length - groups[a].length) || (order.indexOf(groups[a][0]) - order.indexOf(groups[b][0])));

const anchor = (ja) => "tag-" + ja.replace(/^#+/, "");
const enAttr = (ja, en) => (en && en !== ja ? ` data-en="${esc(en)}"` : "");

// タグが1つだけのときは、上のタグ並びは出さない（押す先が自分しかない）
const cloud = ordered.length > 1
  ? ordered.map((t) => `<a href="#${anchor(t)}"${enAttr(t, enOf[t])}>${esc(t)}</a>`).join("")
  : "";
const cloudHtml = cloud ? `  <nav class="tag-cloud" aria-label="タグ" data-en-aria-label="Tags">${cloud}</nav>\n` : "";

const out = ordered.map((t) => {
  const items = groups[t].map((s) => {
    const r = rows[s];
    return `<a class="news-list-row" href="/${s}/"><time datetime="${r.iso}">${r.dot}</time><span class="category${r.mod}"${r.cattr}>${r.cat}</span><strong${r.tattr}>${r.title}</strong></a>`;
  }).join("");
  const n = groups[t].length;
  return `  <section class="tag-group" id="${anchor(t)}">\n` +
    `    <h2${enAttr(t, enOf[t])}>${esc(t)}</h2>\n` +
    `    <p class="tag-count" data-en="${n} post${n > 1 ? "s" : ""}">${n}本</p>\n` +
    `    <div class="news-list">${items}</div>\n` +
    `  </section>\n`;
}).join("");

const pagePath = path.join(ROOT, "tags", "index.html");
const page = readLF(pagePath);
const re = /(<p class="lead"[^>]*>[^\n]*<\/p>\n)[\s\S]*?(<\/main>)/;
if (!re.test(page.text)) { console.error("[中止] tags/index.html の差し替え位置（<p class=\"lead\">〜</main>）が見つからない"); process.exit(1); }
writeEOL(pagePath, page.text.replace(re, (_, head, tail) => head + cloudHtml + out + tail), page.crlf);
console.log(`tags/index.html を作りました（タグ ${ordered.length}／記事 ${Object.keys(tags).length}）`);
